File: NEST_BE/monolith/src/main/java/com/nest/app/fees/entity/FeeType.java
package com.nest.app.fees.entity;

import jakarta.persistence.Column;
import jakarta.persistence.Entity;
import jakarta.persistence.EnumType;
import jakarta.persistence.Enumerated;
import jakarta.persistence.GeneratedValue;
import jakarta.persistence.Id;
import jakarta.persistence.Index;
import jakarta.persistence.Table;
import jakarta.persistence.UniqueConstraint;
import lombok.AllArgsConstructor;
import lombok.Builder;
import lombok.Getter;
import lombok.NoArgsConstructor;
import lombok.Setter;
import org.hibernate.annotations.CreationTimestamp;

import java.math.BigDecimal;
import java.time.Instant;
import java.time.LocalDate;
import java.util.UUID;

/**
 * A named charge an academy raises outside the regular course fee - costume, exam, annual day.
 *
 * <p>Scoped to one academy, never shared across the platform. Two academies both defining a
 * "Costume Fee" is the expected case and they are unrelated records with unrelated amounts, so
 * the uniqueness that matters is {@code (academy_id, name)} and never name on its own.</p>
 */
@Entity
@Table(name = "fee_types",
        uniqueConstraints = @UniqueConstraint(name = "uq_fee_types_academy_name",
                columnNames = {"academy_id", "name"}),
        indexes = @Index(name = "idx_fee_types_academy", columnList = "academy_id"))
@Getter
@Setter
@NoArgsConstructor
@AllArgsConstructor
@Builder
public class FeeType {

    @Id
    @GeneratedValue
    private UUID id;

    @Column(name = "academy_id", nullable = false)
    private UUID academyId;

    @Column(nullable = false, length = 120)
    private String name;

    @Column(nullable = false, precision = 12, scale = 2)
    private BigDecimal amount;

    /**
     * Drives the derived "due" status - an unpaid fee past this date reads as overdue. Null for an
     * open-ended charge with no last date to pay.
     */
    @Column(name = "due_date")
    private LocalDate dueDate;

    /** Pre-selects the mode in the payment recorder. A default, not a restriction. */
    @Enumerated(EnumType.STRING)
    @Column(name = "default_mode", length = 20)
    private FeeMode defaultMode;

    /**
     * Soft-retire rather than delete. A fee type with ledger history behind it has to keep
     * resolving to a name, or every past transaction row loses its label.
     */
    @Column(nullable = false)
    @Builder.Default
    private boolean active = true;

    @Column(name = "created_by")
    private UUID createdBy;

    @CreationTimestamp
    @Column(name = "created_at", nullable = false, updatable = false)
    private Instant createdAt;
}

File: NEST_FE/lib/features/fees/data/fee_type.dart
/// Which batch a fee type applies to, with names so the selector needs no second round trip.
class FeeTypeBatchBinding {
  const FeeTypeBatchBinding({
    required this.batchId,
    required this.batchName,
    required this.courseId,
    required this.courseName,
  });

  final String batchId;
  final String batchName;
  final String courseId;
  final String courseName;

  /// How the prototype labels a batch in the Other Fees selector: "Guitar Beginner · Batch A".
  /// A batch name alone is ambiguous - every course has a "Batch A".
  String get label => '$courseName · $batchName';

  factory FeeTypeBatchBinding.fromJson(Map<String, dynamic> json) => FeeTypeBatchBinding(
        batchId: json['batchId'] as String,
        batchName: json['batchName'] as String? ?? 'Batch',
        courseId: json['courseId'] as String? ?? '',
        courseName: json['courseName'] as String? ?? 'Unknown course',
      );
}

/// A named charge raised outside the regular course fee - costume, exam, annual day.
class FeeType {
  const FeeType({
    required this.id,
    required this.name,
    required this.amount,
    required this.batches,
    this.dueDate,
    this.defaultMode,
    this.active = true,
  });

  final String id;
  final String name;
  final num amount;

  /// The last date to pay. An unpaid fee past it reads as overdue. Null for an open-ended charge.
  final DateTime? dueDate;

  final String? defaultMode;
  final bool active;
  final List<FeeTypeBatchBinding> batches;

  /// A fee bound to exactly one batch is not a choice - the selector locks to it rather than
  /// making the admin open a menu with a single entry.
  FeeTypeBatchBinding? get onlyBatch => batches.length == 1 ? batches.first : null;

  factory FeeType.fromJson(Map<String, dynamic> json) => FeeType(
        id: json['id'] as String,
        name: json['name'] as String? ?? 'Fee',
        amount: json['amount'] as num? ?? 0,
        dueDate: DateTime.tryParse(json['dueDate'] as String? ?? ''),
        defaultMode: json['defaultMode'] as String?,
        active: json['active'] as bool? ?? true,
        batches: (json['batches'] as List<dynamic>? ?? [])
            .map((e) => FeeTypeBatchBinding.fromJson(e as Map<String, dynamic>))
            .toList(),
      );
}
